dbms/src/TiDB/Raft/RaftLogManager.cpp

```diff
@@ -0,0 +1,184 @@
+                = std::min(hint_iter->second.eager_truncate_index, eager_truncated_index);
+            hint_iter->second.applied_index = std::max(hint_iter->second.applied_index, applied_index);
+            return true;
+        }
+        // No much updates after last GC task is built, let's just keep the task hint unchanged
+        return false;
+    }
+
+    // New GC task hint is created
+    // LOG_DEBUG(Logger::get(), "create hint, region_id={} first_index={} applied_index={}", region_id, first_index, applied_index);
+    tasks[region_id] = RaftLogEagerGcHint{
+        .eager_truncate_index = eager_truncated_index,
+        .applied_index = applied_index,
+    };
+    return true;
+}
+
+RaftLogEagerGcTasks::Hints RaftLogEagerGcTasks::getAndClearHints()
+{
+    std::unique_lock lock(mtx_tasks);
+    Hints hints;
+    hints.swap(tasks);
+    return hints;
+}
+
+struct RegionGcTask
+{
+    bool skip;
+    UInt64 eager_truncate_index;
+    UInt64 applied_index;
+};
+
+RegionGcTask getRegionTask(
+    const RegionID region_id,
+    const raft_serverpb::RaftApplyState & region_state,
+    const RaftLogEagerGcHint & hint,
+    UInt64 threshold,
+    const LoggerPtr & logger)
+{
+    // LOG_DEBUG(logger, "Load persisted region apply state, region_id={} state={}", region_id, region_state.ShortDebugString());
+    if (region_state.applied_index() <= 1)
+        return RegionGcTask{.skip = true, .eager_truncate_index = 0, .applied_index = 0};
+
+    RegionGcTask task{.skip = false, .eager_truncate_index = 0, .applied_index = 0};
+    task.eager_truncate_index = hint.eager_truncate_index;
+    task.applied_index = region_state.applied_index() - 1;
+    if (task.applied_index <= task.eager_truncate_index || task.applied_index - task.eager_truncate_index <= threshold)
+    {
+        LOG_DEBUG(
+            logger,
+            "Skip eager gc, region_id={} persist_first_index={} persist_applied_index={} hint_eager_truncate_index={} "
+            "hint_applied_index={}",
+            region_id,
+            region_state.truncated_state().index(),
+            region_state.applied_index(),
+            hint.eager_truncate_index,
+            hint.applied_index);
+        task.skip = true;
+    }
+    return task;
+}
+
+RaftLogGcTasksRes executeRaftLogGcTasks(Context & global_ctx, RaftLogEagerGcTasks::Hints && hints)
+{
+    auto write_node_ps = global_ctx.tryGetWriteNodePageStorage();
+    if (!write_node_ps)
+        return {};
+    assert(write_node_ps);
+
+    const auto eager_gc_rows = global_ctx.getTMTContext().getKVStore()->getRaftLogEagerGCRows();
+    LoggerPtr logger = Logger::get();
+
+    Stopwatch watch;
+    RaftLogGcTasksRes eager_truncated_indexes;
+    size_t num_skip = 0;
+    size_t total_num_raft_log_removed = 0;
+    for (const auto & [region_id, hint] : hints)
+    {
+        // Read region apply_state from UniPS
+        auto region_apply_state_key = UniversalPageIdFormat::toRaftApplyStateKeyInKVEngine(region_id);
```

Comment:
`Region::lastCompactLogApplied` will be updated to current `applied_index` every time we finished a flush in `persistRegion`.
And will be updated to what is read from `RegionPersister` after restart.
I think may be we can just use what `Region::lastCompactLogApplied` returns, instead of performing a extra read?
Still, we need to assert `Region::lastCompactLogApplied` to be not zero here.

Reply:
1. There is a time gap between we update `Region::lastCompactLogApplied` and tiflash-proxy real persist the region apply state.
2. We must ensure all needed RaftLog can be read after restart

So we have to read the region apply state from UniPS to ensure the correctness

I've updated the way to read RegionApplyState in UniPS

Reply:
You are right, may be we can put these as comments for these codes?

Reply:
I've added some comments here. Or you think we should add comments about why not using the member of `Region` in TiFlash memory?
https://github.com/pingcap/tiflash/blob/2904e72374e1e36b454d8164d54114f626614cad/dbms/src/Storages/Transaction/RaftLogManager.cpp#L136-L143